Threading API conversion, vol. 2

ghedo opened the pull request:
A few more conversions, we are almost there!!

But the "big ones" (`md_rand,c`, `mem_dbg.c`, `engines/`, RSA blinding) still need conversion and will likely take some more time (not sure if I can do this in time for the freeze, help would be appreciated :) ).

ghedo commented:
Interestingly there are a bunch of locks that are currently unused:
```
# define CRYPTO_LOCK_ECDSA               32
# define CRYPTO_LOCK_ECDH                34
# define CRYPTO_LOCK_BN                  35
# define CRYPTO_LOCK_STORE               37
# define CRYPTO_LOCK_COMP                38
# define CRYPTO_LOCK_FIPS                39
# define CRYPTO_LOCK_FIPS2               40
```

ajmohan commented on crypto/bio/b_addr.c:
```diff
@@ -55,12 +55,16 @@
 #include <ctype.h>
 
+CRYPTO_RWLOCK *bio_lookup_lock;
+CRYPTO_ONCE bio_lookup_init = CRYPTO_ONCE_STATIC_INIT;
```
both may be defined as static.

ghedo replied:
Fixed, thanks.

mattcaswell commented:
+1

mattcaswell commented:
@ghedo, have you started: md_rand,c, mem_dbg.c, engines/, RSA blinding? I can maybe take some of this on.

mattcaswell commented:
@ghedo, I took a gamble and I have a patch for md_rand.c

ghedo commented:
@mattcaswell I only looked at RSA blinding but I haven't finished yet. I sort of wanted to convert that code to something like what BoringSSL does (which is saner IMO) but I guess it's too late for that at this point:
https://boringssl.googlesource.com/boringssl/+/master/crypto/rsa/rsa_impl.c#212

ghedo commented:
I just fixed a couple of memory leaks: the RSA lock wasn't freed properly and I implemented the `CRYPTO_MEM_CHECK_DISABLE` dance for `ex_data_lock` (I don't think we ever want to free that, but maybe the init code could be used for this? I don't really know how that works).

mattcaswell commented:
Still +1...ping @richsalz.

richsalz commented:
will merge.  @mattcaswell you should post your branch here for @ghedo to look at.

mattcaswell commented:
You mean the one you just +1'd?? How about I just merge it :-)

richsalz commented:
Please do ☺
Feel free to merge anything that’s ready, I was just picking up the slack for others with more important stuff.

richsalz commented:
sorry, not enough email context.  I meant that @ghedo might have useful feedback on what I just plus-one'd.

And I merged this and am closing it. :)
